src/org/mozilla/javascript/optimizer/OptRuntime.java

```diff
@@ -22,116 +20,88 @@
+    public static Object callN(
+            Callable fun, Scriptable thisObj, Object[] args, Context cx, Scriptable scope) {
         return fun.call(cx, scope, thisObj, args);
     }
 
-    /**
-     * Implement name(args) call shrinking optimizer code.
-     */
-    public static Object callName(Object[] args, String name,
-                                  Context cx, Scriptable scope)
-    {
+    /** Implement name(args) call shrinking optimizer code. */
+    public static Object callName(Object[] args, String name, Context cx, Scriptable scope) {
         Callable f = getNameFunctionAndThis(name, cx, scope);
         Scriptable thisObj = lastStoredScriptable(cx);
         return f.call(cx, scope, thisObj, args);
     }
 
-    /**
-     * Implement name() call shrinking optimizer code.
-     */
-    public static Object callName0(String name,
-                                   Context cx, Scriptable scope)
-    {
+    /** Implement name() call shrinking optimizer code. */
+    public static Object callName0(String name, Context cx, Scriptable scope) {
         Callable f = getNameFunctionAndThis(name, cx, scope);
         Scriptable thisObj = lastStoredScriptable(cx);
         return f.call(cx, scope, thisObj, ScriptRuntime.emptyArgs);
     }
 
-    /**
-     * Implement x.property() call shrinking optimizer code.
-     */
-    public static Object callProp0(Object value, String property,
-                                   Context cx, Scriptable scope)
-    {
+    /** Implement x.property() call shrinking optimizer code. */
+    public static Object callProp0(Object value, String property, Context cx, Scriptable scope) {
         Callable f = getPropFunctionAndThis(value, property, cx, scope);
         Scriptable thisObj = lastStoredScriptable(cx);
         return f.call(cx, scope, thisObj, ScriptRuntime.emptyArgs);
     }
 
-    public static Object add(Object val1, double val2)
-    {
-        if (val1 instanceof Scriptable)
-            val1 = ((Scriptable) val1).getDefaultValue(null);
-        if (!(val1 instanceof CharSequence))
-            return wrapDouble(toNumber(val1) + val2);
-        return new ConsString((CharSequence)val1, toString(val2));
+    public static Object add(Object val1, double val2, Context cx) {
+        if (val1 instanceof Integer) {
+            return ((Integer) val1) + val2;
+        }
+        if (val1 instanceof Double) {
+            return ((Double) val1) + val2;
+        }
+        return ScriptRuntime.add(val1, val2, cx);
     }
 
-    public static Object add(double val1, Object val2)
-    {
-        if (val2 instanceof Scriptable)
-            val2 = ((Scriptable) val2).getDefaultValue(null);
-        if (!(val2 instanceof CharSequence))
-            return wrapDouble(toNumber(val2) + val1);
-        return new ConsString(toString(val1), (CharSequence)val2);
+    public static Object add(double val1, Object val2, Context cx) {
+        if (val2 instanceof Integer) {
```

Comment:
I think `java.lang.Double` is used more than `java.lang.Integer` inside Rhino.
I grep with `wrapInt`, I found 14 matches, but with `wrapNumber`, I found 90 matches.

In the future, do you want to use integers like v8 to make it faster?
https://medium.com/fhinkel/v8-internals-how-small-is-a-small-integer-e0badc18b6da

Reply:
Yes, in general I'd like to try and start steering Rhino into using Integers more when we can. (For example, the standard constant for "zero" is a double, which means that we use doubles in every "for (var i = 0; i < foo; i++)" loop!

I agree that doubles are more common and I'll try switching them. 

Over time I want to go back to using invokedynamic in the bytecode for all the super-complex operations in ScriptRuntime like "add," "compare," "eq," etc. That will allow the JVM at runtime to re-order which operations are most likely to happen at a particular call site. However I still haven't come up with a single place where this makes a big difference yet, as opposed to simpler optimizations like this one.